File: Baekjoon/Problem_1000~1999/Problem_1103/main.py
import sys
sys.setrecursionlimit(10 ** 6)

input = lambda: sys.stdin.readline().rstrip()
is_in = lambda r, c: (0 < r <= N) and (0 < c <= M)

dr = [ 0, -1, 0, 1 ]
dc = [ 1, 0, -1, 0 ]

def dfs(r0: int=1, c0: int=1) -> int:
    global visited, n_descendants

    w = ord(board[r0][c0]) - ord('0')
    for i in range(4):
        r, c = r0 + w * dr[i], c0 + w * dc[i]
        
        if not is_in(r, c): continue # 범위 밖으로 벗어난 경우
        elif board[r][c] == 'H': continue # 다음 위치가 hole인 경우
        elif visited[r][c]: return -1 # loop를 식별한 경우
        elif n_descendants[r][c] != 0 and n_descendants[r0][c0] > n_descendants[r][c]: continue # 해당 위치로 이동하는 것이 오히려 부족한 경우

        visited[r][c] = True
        n_child_descendants = dfs(r, c)
        visited[r][c] = False
        if n_child_descendants == -1: return -1
        
        n_descendants[r0][c0] = max(n_descendants[r0][c0], n_child_descendants)
        
    return n_descendants[r0][c0] + 1

if __name__ == "__main__":
    N, M = map(int, input().split())
    board = [ '0' * (M+1) ]
    
    for r in range(N): board.append('0' + input())
    
    visited = [ [ False ] * (M+1) for _ in range(N+1) ] 
    n_descendants = [ [ 0 ] * (M+1) for _ in range(N+1) ] # 현재 위치에서 갈 수 있는 최대 깊이...
    visited[1][1] = True
    print(dfs())

File: Baekjoon/Problem_1000~1999/Problem_1103/main.cpp
#include <cstdio>
#include <iostream>
#include <algorithm>

using namespace std;
const int SIZE = 51;
const int dr[4] = { -1, 1, 0, 0 };
const int dc[4] = { 0, 0, -1, 1 };

int N, M;
char board[SIZE][SIZE];
bool visited[SIZE][SIZE];
int n_descendants[SIZE][SIZE];

inline bool is_in(const int r, const int c) {
    return ((0 < r) && (r <= N)) && ((0 < c) && (c <= M));
}

int dfs(const int r0 = 1, const int c0 = 1) {
    int w = board[r0][c0] - '0';
    for (int i = 0; i < 4; i++) {
        int r = r0 + w * dr[i], c = c0 + w * dc[i];

        if (!is_in(r, c)) continue;
        else if (board[r][c] == 'H') continue;
        else if (visited[r][c]) return -1;
        else if (n_descendants[r][c] != 0 && n_descendants[r0][c0] > n_descendants[r][c]) continue;

        visited[r][c] = true;
        int ret = dfs(r, c);
        visited[r][c] = false;
        if (ret == -1) return -1;

        n_descendants[r0][c0] = max(n_descendants[r0][c0], ret);
    }
    return n_descendants[r0][c0] + 1;
}

int main(void) {
    scanf("%d %d", &N, &M);
    for (int r = 1; r <= N; r++) {
        for (int c = 1; c <= M; c++) scanf(" %1c", &board[r][c]);
    }

    visited[1][1] = true;
    printf("%d\n", dfs());

    return 0;
}